### PR TITLE
Add large data set support to CLM channel selection

### PR DESCRIPTION
**Pull request by Etheryte**
## What does this PR change?

Update the CLM channels selection view to not lock up on large data sets.  

Main points:  
 - reduce unnecessary rerenders with memoing and update checks  
 - replace simple list rendering with a virtualized list implementation (https://github.com/petyosi/react-virtuoso)
 - reduce computational complexity of some of the data transforms

## GUI diff

Perhaps some small layout shifts here and there, but should look more or less the same.

- [x] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [x] **DONE**

## Test coverage
- No tests: already covered

- [x] **DONE**

## Links

Part of https://bugzilla.suse.com/show_bug.cgi?id=1192598

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: web/html/src/manager/content-management/shared/components/panels/sources/channels/channels-api.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                recommended: false,
+                standardizedName: "",
+                parent: base,
+              },
+            ],
+          });
+        }
+      }
+
+      // TODO: This matches the current behavior but this should already be done on the server side, with a separate flag if necessary
```
Comment (Etheryte) [inline web/html/src/manager/content-management/shared/components/panels/sources/channels/channels-api.ts:102]: @cbbayburt Is this something we can move to the server, or is there a reason why not to do it that I'm missing?

**Review by cbbayburt — COMMENTED** [review]
Thanks, @Etheryte, looks very clean

One thing from me: let's try and address all TODOs as much as we can now, while it's fresh. Otherwise, they'll age here forever.
My suggestion is to;
1. Do all TODOs that can be done now in a reasonable time
2. Reach out if they need to be unblocked by someone else
3. If cannot be done now, but it's important; make a plan (describing it in a separate issue)
4. If none of the above is feasible, just delete the TODO

File: web/html/src/manager/content-management/project/project.tsx [hunk 2]
```diff
@@ -41,8 +42,11 @@ const Project = (props: Props) => {
   const hasEditingPermissions = isOrgAdmin(roles);
 
   useInterval(() => {
-    onAction({}, "get", project.properties.label).then((project) => {
-      setProject(project);
+    onAction({}, "get", project.properties.label).then((newProject) => {
+      // Only propagate the received value if something has changed
+      if (!_isEqual(project, newProject)) {
+        setProject(newProject);
+      }
```
[the remark this answers is not in the document]
Comment (Etheryte) [reply web/html/src/manager/content-management/project/project.tsx:49]: This is a part of the issue at hand actually, the view was pretty heavy to render and on top of that it got rerendered whenever the recurring network request replied even if the contents were the same as before.

File: web/html/src/core/channels/type/channels.type.ts [hunk 3]
```diff
@@ -9,3 +11,62 @@ export type ChannelType = {
...
+  id: number;
+  archLabel: string;
+  custom: boolean;
+  isCloned: boolean;
+  label: string;
+  name: string;
+  standardizedName: string;
+  recommended: boolean;
+  subscribable: boolean;
+};
```
[the remark this answers is not in the document]
Comment (cbbayburt) [reply web/html/src/core/channels/type/channels.type.ts]: Oops sorry, I was looking at the 1st commit alone

File: web/html/src/manager/content-management/shared/components/panels/sources/channels/channels-api.ts (hunk 1 above)
Comment (Etheryte) [reply web/html/src/manager/content-management/shared/components/panels/sources/channels/channels-api.ts]: I've made the change in[`ChannelsApiController.java`](https://github.com/uyuni-project/uyuni/pull/4739/files#diff-4d83abbd4ca969f44973a005de0adc5886dd4772d704886ab010eff4008c8280), if you have a moment to check that my copy-paste-from-google Java coding makes sense, I'd appreciate it. 😄

File: java/code/src/com/suse/manager/webui/controllers/channels/ChannelsApiController.java [hunk 4]
```diff
@@ -77,6 +79,12 @@ public static String getAllChannels(Request req, Response res, User user) {
                             .filter(c -> !filterOutIds.contains(c.getBase().getId()))
                             .collect(Collectors.toList()
             );
+            Collections.sort(jsonChannelsFiltered, new Comparator<ChannelsJson>() {
+                @Override
+                public int compare(ChannelsJson c1, ChannelsJson c2) {
+                    return c1.getBase().getId().compareTo(c2.getBase().getId());
+                }
+            });
```
Comment (cbbayburt) [inline java/code/src/com/suse/manager/webui/controllers/channels/ChannelsApiController.java:87]: It works just fine, but here's a simpler one-liner:
```suggestion
            jsonChannelsFiltered.sort(Comparator.comparing(cIn -> cIn.getBase().getId()));
```

Comment (Etheryte) [reply java/code/src/com/suse/manager/webui/controllers/channels/ChannelsApiController.java]: Nice, thanks a lot for this. I don't know enough about Java so I wouldn't even have thought of something like this. 🍻

**Pull request merged**